### DISCUSSION TITLE
Can ipxe support AX88179A & RTL8153?

### DISCUSSION THREAD

**Discussion by wtt101** [ipxe/ipxe#685] [Q&A]
The system i currently using require to load the driver before it can execute ipxe.efi file. It will hit DHCP error without loading the driver

Both of these are USB NIC Adapter 
USB31000S - AX88179A 
UE300 - RTL8153

**Comment by NiKiZe** [comment]
Do you have the usbids for those devices?
I remember having previous discussions about AX88179 which has a native driver in axge.c but A is different(?)

Most usb ethernet devices also support ncm or ecm, so you could build an axge+acm+ncm+ecm.usb or similar and test?

**Reply by stappersg** [reply]
```text
stappers@myhost:~/src/mailinglists/ipxe/src
$ make bin/ache+acm+ncm+ecm.usb
  [AR] bin/blib.a
ar: creating bin/blib.a
  [BIN] bin/usbdisk.bin
  [VERSION] bin/version.ache+acm+ncm+ecm.hd.o
  [LD] bin/ache+acm+ncm+ecm.hd.tmp
ld:--defsym:2: syntax error
make: *** [Makefile.housekeeping:1233: bin/ache+acm+ncm+ecm.hd.tmp] Error 1
rm bin/version.ache+acm+ncm+ecm.hd.o bin/usbdisk.bin
stappers@myhost:~/src/mailinglists/ipxe/src
$ make bin/ncm+ecm.usb
  [BIN] bin/usbdisk.bin
  [VERSION] bin/version.ncm+ecm.hd.o
  [LD] bin/ncm+ecm.hd.tmp
ld:--defsym:2: syntax error
make: *** [Makefile.housekeeping:1233: bin/ncm+ecm.hd.tmp] Error 1
rm bin/version.ncm+ecm.hd.o bin/usbdisk.bin
stappers@myhost:~/src/mailinglists/ipxe/src
$ make bin/ncm.usb
  [BIN] bin/usbdisk.bin
  [VERSION] bin/version.ncm.hd.o
  [LD] bin/ncm.hd.tmp
  [BIN] bin/ncm.hd.bin
  [ZINFO] bin/ncm.hd.zinfo
  [ZBIN] bin/ncm.hd.zbin
  [FINISH] bin/ncm.hd
  [FINISH] bin/ncm.usb
rm bin/version.ncm.hd.o bin/usbdisk.bin bin/ncm.hd.zinfo bin/ncm.hd.zbin bin/ncm.hd bin/ncm.hd.bin
stappers@myhost:~/src/mailinglists/ipxe/src
$ file bin/ncm.usb
bin/ncm.usb: DOS/MBR boot sector; partition 4 : ID=0xeb, active, start-CHS (0x0,1,1), end-CHS (0x1,63,32), startsector 32, 4064 sectors
stappers@myhost:~/src/mailinglists/ipxe/src
$ 
```

**Reply by wtt101** [reply]
Could you guide me on how to build it? Currently, i'm using FOG Project with iPXE, currently im using snp.efi file and it able to support onboard & pcie NIC but snp.efi cant support usb nic.

After you share on "ncm or ecm" i tried ncm--ecm--axge.efi and with this it support for the USB NIC adapter. Can you guide me on how can i consolidate both snp.efi and ncm--ecm--axge.efi together ?

**Build Command by FOG Project**
https://github.com/FOGProject/fogproject/blob/master/utils/FOGiPXE/buildipxe.sh
**EFI files**
https://github.com/FOGProject/fogproject/tree/master/packages/tftp

May i know how can i get the usbids?

**Comment by stappersg** [comment]
On Tue, Jul 19, 2022 at 08:13:20AM -0700, wtt101 wrote:
> 
> May i know how can i get the usbids?
> 

On Linux:   lsusb

**Comment by wtt101** [comment]
Below are the usbids. Can you guide me on how to build the ipxe rom for below USB NIC so i can replace the existing one?
USB31000S (StarTech) - AX88179A | 0b95:1790
UE300(TP-LINK) – RTL8153 | 2357:0601

**Reply by stappersg** [reply]
Nofitication email that I got:
```text
Date: Tue, 19 Jul 2022 22:46:44 -0700
From: weitheng10 <notifications@github.com>
To: ipxe/ipxe <ipxe@noreply.github.com>
Cc: Geert Stappers <stappers@stappers.it>, Comment <comment@noreply.github.com>
Subject: Re: [ipxe/ipxe] Can ipxe support AX88179A & RTL8153? (Discussion #685)

My usb nic id is 2357:0601

--  
Reply to this email directly or view it on GitHub:
https://github.com/ipxe/ipxe/discussions/685#discussioncomment-3186164
You are receiving this because you commented.

Message ID: <ipxe/ipxe/repo-discussions/685/comments/3186164@github.com>
```

Result of my research triggered by a _"hey, that is strange"_
* [@weitheng10](https://github.com/weitheng10) is a contributor of the FOGproject.
* [@wtt101](https://github.com/wtt101) is an account without prior reputation.


I think that the @wtt101 account that doesn't want to build a further reputation.

@weitheng10 Thank you for inflating the value of a github account.

**Comment by stappersg** [comment]
```text
stappers@myhost:~/src/mailinglists/ipxe/src
$ make bin/axge--acm--ncm--ecm.usb
  [PARSEROM]
  [AR] bin/blib.a
ar: creating bin/blib.a
  [BIN] bin/usbdisk.bin
  [VERSION] bin/version.axge--acm--ncm--ecm.hd.o
  [LD] bin/axge--acm--ncm--ecm.hd.tmp
  [BIN] bin/axge--acm--ncm--ecm.hd.bin
  [ZINFO] bin/axge--acm--ncm--ecm.hd.zinfo
  [ZBIN] bin/axge--acm--ncm--ecm.hd.zbin
  [FINISH] bin/axge--acm--ncm--ecm.hd
  [FINISH] bin/axge--acm--ncm--ecm.usb
rm bin/version.axge--acm--ncm--ecm.hd.o bin/axge--acm--ncm--ecm.hd.bin \
 bin/axge--acm--ncm--ecm.hd bin/usbdisk.bin bin/axge--acm--ncm--ecm.hd.zbin bin/axge--acm--ncm--ecm.hd.zinfo
stappers@myhost:~/src/mailinglists/ipxe/src
$ make bin/axge--acm--ncm--ecm.usb
make: 'bin/axge--acm--ncm--ecm.usb' is up-to-date.
stappers@myhost:~/src/mailinglists/ipxe/src
$ 
```
and retransmitting:
> Most usb ethernet devices also support ncm or ecm, so you could build an axge+acm+ncm+ecm.usb or similar and test?

**Comment by wtt101** [comment]
I have tested with axge--acm--ncm--ecm.efi build it able to proceed to iPXE menu without any error but:
 If I use "Windows Button > Restart" to reboot my machine and execute axge--acm--ncm--ecm.efi i will face below error
![1](https://user-images.githubusercontent.com/109604840/180936102-e29d2a84-7830-4046-894e-accceba5c427.png)

If i manually power off and on the machine then iPXE menu working as expected without error.

This happen to Windows 10 when im using Linux i don't face any error. May I know how can i troubleshoot this?

**Comment by stappersg** [comment]
On Mon, Jul 25, 2022 at 11:14:25PM -0700, wtt101 wrote:
> 
> If i manually power off and on the machine then iPXE menu working as expected without error.
> 

And that is with each of the devices?
In case it is only one: Name the working one.

**Reply by wtt101** [reply]
To clarify, 

Scenario 1 : iPXE Success | Windows 10
Machine is Off > Power On > execute startup.nsh (will call axge--acm--ncm--ecm.efi) > iPXE Menu > Boot into Windows 10

Scenario 2 : iPXE Success | Windows 10 
Machine booted into Windows 10 > Power Off (Cut Off Electric) > Power On > execute startup.nsh (will call axge--acm--ncm--ecm.efi) > iPXE Menu > Boot into Windows 10

Scenario 3 : iPXE Failed | Windows 10 
Machine booted into Windows 10 > Press Windows + Restart > System Reboot > execute startup.nsh (will call axge--acm--ncm--ecm.efi) > iPXE (No Configuration Method Succeeded Press... DHCP Fail press any key to reboot > System Reboot >it will be in loop

#to exit the reboot loop, i need to power off and on

Scenario 4 : iPXE Success | Linux Debian
Machine booted into Debian > input reboot command > System Reboot > execute startup.nsh (will call axge--acm--ncm--ecm.efi) > iPXE Menu > Boot into Debian

I'm using same machine, same usb NIC, same efi file. Only different is OS. 
May I know how can i troubleshoot the root cause of scenario 3 failing?

**Reply by stappersg** [reply]
And that is with each of the devices?
In case it is only one: Name the working one.
